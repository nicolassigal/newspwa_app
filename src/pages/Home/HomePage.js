import React, { Component } from "react";
import ArticleList from '../../components/ArticleList';
import NoArticles from './../../components/NoArticles';
class HomePage extends Component {
    constructor(props) {
        super(props);

        this.state = {
            news: [],
            page: 1,
            loading: false
        };

        this.fetchNews = this.fetchNews.bind(this);

        window.addEventListener('scroll', this.scrollHandler.bind(this));
    }

    async componentDidMount() {
        await this.fetchNews();
    }

    async fetchNews() {
        let {page, news: oldNews, loading} = this.state;
        this.setState({loading: true})
        try {
        const response = await fetch(`${API_URL}/top-headlines?country=us&page=${page}`);
        const newsResp = await response.json();
        if (newsResp.length) {
            const news = [...oldNews, ...newsResp];
            page = page + 1;
            this.setState({ news, page, loading:false });
        }
        } catch(e) {
            this.setState({ loading:false });
            console.log("failed to fetch home news");
        }
    }

    async scrollHandler() {
        const scrollPos = window.innerHeight + document.documentElement.scrollTop;
        const offset = document.documentElement.offsetHeight - 800;
        if (scrollPos > offset  && !this.state.loading) {
            await this.fetchNews();
        }
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.scrollHandler);
    }

    render() {
        const { news, loading } = this.state;
        return (
            <div className="home-page">
                {!loading && !news.length && <NoArticles />}
                {news.length ? <ArticleList articles={news} /> : null }
            </div>
        );
    }
}

export default HomePage;
