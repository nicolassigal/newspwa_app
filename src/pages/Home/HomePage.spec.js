import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import HomePage from './HomePage';

describe('Home Page', () => {
  it('Home page should render correctly', () => {
    const tree = renderer
      .create(<HomePage />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});