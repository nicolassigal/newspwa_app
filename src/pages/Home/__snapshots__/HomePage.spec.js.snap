// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Home page should render correctly 1`] = `
<div
  className="home-page"
>
  Home Page
</div>
`;
